### DISCUSSION TITLE
Cannot reproduce the experiment result in the paper

### DISCUSSION THREAD

**Discussion by dxyzx0** [ds4dm/ecole#262] [Questions]
My SCIP version: `7.0.3`.

I use the notebook `branching-imitation.ipynb` to reproduce the results in paper and try to change `MAX_SAMPLES` from `1000` to `120000` and `350000`. 

Except for `Capacitated Facility Location`, which works well on all 3 categories, we can only train model on `easy` instances( and generalize them to other `easy` instances) but cannot generalize to `medium` and `hard`. 

There are some steps I have tried and problems I have encountered:

1.  Combinatorial Auction training accuracy doesn't increase like other problems. I tried to adjust `LEARNING_RATE` from `0.0001` to `0.01`, `0.001` and `0.00001` and `emb_size` from `64` to `128`, but none of them increase the training/valid accuracy. But the model has little positive effect when generalized to `easy` instances.
<img src="https://user-images.githubusercontent.com/28868457/135940947-9042198f-faac-4246-bab2-8e60d4650051.png" width="200">

2. Combinatorial Auction / Set Cover/ Maximum Independent Set can generalize on `easy` instances but not on `medium` and `hard` instances. For `easy` instances, they all have little positive effect but on `medium` and `hard` instances, they all have negative effect.
Combinatorial Auction:
<img src="https://user-images.githubusercontent.com/28868457/135941270-e18d8da2-0108-4dfd-a6f5-27bc0987789e.png" width="200">

Independent set:
<img src="https://user-images.githubusercontent.com/28868457/135941247-f27a1515-6df4-4156-8b74-116dfd99270c.png" width="200">

Set cover:
<img src="https://user-images.githubusercontent.com/28868457/135941214-cce94143-5f92-488e-808c-3ae64ed70c3e.png" width="200">

3. Capacitated Facility Location works well on all 3 categories, but SCIP works better on hard problems than in the paper:
<img src="https://user-images.githubusercontent.com/28868457/135941988-1ee1cda8-9407-4caf-8474-75bca707c7fd.png" width="200">
<img src="https://user-images.githubusercontent.com/28868457/135942065-4c18c8a1-d78b-4671-b90c-1d40ce79f64a.png" width="600">

Any advice on reproducing the result?

**Comment by dchetelat** [comment]
Hi, I'm going to have a look, this tutorial was based on (unreleased) code with Ecole that was definitely working.

Perhaps something changed in the latest version that broke the code reproduction? I'll check and get back to you.

**Reply by dxyzx0** [reply]
I can send you my training and validation log.

That's contents of my zip:

I put `logs` and `trained-params.pkl` for each task in their own folder and provide my reproducing `setcover` task code for example.

In `xxx_eval.log`, the first section is the evaluation result for `easy`, second for `medium` and third for `hard`. `1003_xxx.log` for data size `350000` and `0930_xxx.log` for data size `120000`.

In folder `code`, It's only for task `setcover`.

In `result.docx`, I summarize the `training` and `validation` result for each task.

You can access the files here:

> transfer link: https://cowtransfer.com/s/6f7781369cfd45 or access [ cowtransfer.com ] and input extract code: anqwwd to download;

**Reply by dchetelat** [reply]
Thanks - I'll get onto it. It might take some time, as running these experiments take a while!

**Reply by dxyzx0** [reply]
My evaluation result summary for data-size `350000`.

*   auction：

    *   eval:

        ```
        easy_nodes_scip: 352.26
        easy_nodes_gnn: 390.16
        easy_time_scip: 6.795400000000001
        easy_time_gnn: 5.0126
        medium_nodes_scip: 24122.0
        medium_nodes_gnn: 22906.72
        medium_time_scip: 189.0156
        medium_time_gnn: 279.44559999999996
        hard_nodes_scip: 159922.26315789475
        hard_nodes_gnn: 127835.84210526316
        hard_time_scip: 2267.422631578948
        hard_time_gnn: 2532.1652631578945
        ```

        

*   cfl

    *   eval:

        ```
        easy_nodes_scip: 299.8
        easy_nodes_gnn: 638.1
        easy_time_scip: 63.70600000000002
        easy_time_gnn: 61.836666666666666
        medium_nodes_scip: 515.1
        medium_nodes_gnn: 1058.8333333333333
        medium_time_scip: 272.44199999999995
        medium_time_gnn: 273.7713333333333
        hard_nodes_scip: 260.26666666666665
        hard_nodes_gnn: 734.1333333333333
        hard_time_scip: 735.6453333333334
        hard_time_gnn: 879.7253333333333
        ```

        

*   indset:

    *   eval:

        ```
        easy_nodes_scip: 151.46666666666667
        easy_nodes_gnn: 1358.5333333333333
        easy_time_scip: 9.817
        easy_time_gnn: 15.673
        medium_nodes_scip: 2368.8333333333335
        medium_nodes_gnn: 71228.66666666667
        medium_time_scip: 83.72500000000001
        medium_time_gnn: 951.8249999999999
        ```

        

*   setcover:

    *   eval:

        ```
        easy_nodes_scip: 17.833333333333332
        easy_nodes_gnn: 90.63333333333334
        easy_time_scip: 3.3183333333333342
        easy_time_gnn: 2.661
        medium_nodes_scip: 1192.8
        medium_nodes_gnn: 8520.5
        medium_time_scip: 19.830000000000002
        medium_time_gnn: 88.54366666666667
        hard_nodes_scip: 33628.73333333333
        hard_nodes_gnn: 125083.83333333333
        hard_time_scip: 349.88566666666674
        hard_time_gnn: 1872.7210000000002
        ```

**Reply by dxyzx0** [reply]
Is it possible to open source your unreleased code with ecole?

**Reply by BenjaDomb** [reply]
I am also trying to reproduce the paper results. It would be great if you can open source your unreleased code with ecole

**Reply by dchetelat** [reply]
Yes I think this is the best option. I'll first check whether that code still works with the latest version of Ecole, and after fixing things if necessary, I'll create a public repo.

**Reply by BenjaDomb** [reply]
It would be great!

**Comment by dchetelat** [comment]
Hi @DingXiangYuanZhiXing and @benjadb,

I spent some time cleaning up the internal code, and re-running experiments. Because I didn't have much time, I only ran setcover with seed 0, to get a general idea. To also figure out if the problems were coming from the instance generation or Ecole/your code, I also decided to first start by running the code on the old instances that had been used for the original paper experiments, which I still had.

The results were as follows:
```
small
  internal:relpscost  :  nodes     50.90, node winner   19/20, time     8.28, time winner    0/20
  gcnn:supervised     :  nodes    134.57, node winner    1/20, time     5.36, time winner   20/20
medium
  internal:relpscost  :  nodes   2403.72, node winner    4/20, time    66.97, time winner    0/20
  gcnn:supervised     :  nodes   1938.38, node winner   16/20, time    42.57, time winner   20/20
big
  internal:relpscost  :  nodes  84308.08, node winner    3/13, time  1827.75, time winner    4/13
  gcnn:supervised     :  nodes  64842.46, node winner   11/13, time  1689.49, time winner   10/13
```
Although this was only for one seed and on one dataset, these results seem consistent with those reported in the original paper. So it seems like this code is working, that Ecole didn't break the GNN, and that the issues are either with the instance generation, or your code implementations.

My next steps would be to write code for the instance generation, and re-run this code on the newly generated instances. Then we will know if the issue is your code or the Ecole instance generation. In the meanwhile, I created a new repo with the cleaned-up code. You can find it [here](https://github.com/ds4dm/learn2branch-ecole). It is missing the instance generation code, which I still need to write, run, and debug. If eventually I manage to make it work, I'll add the missing instance generation code and I will publicize the repo on the [original learn2branch repo](https://github.com/ds4dm/learn2branch) as an official reimplementation. In the meanwhile, feel free to use the code with the instances generated by the previous repo, or on other instances of your choice.

**Reply by dxyzx0** [reply]
Thank you very much! I have a few questions:
1. Did you train the model on big instances or small instances? Since the nodes on big are less than small.
2. From [this line](https://github.com/ds4dm/learn2branch-ecole/blob/b1278da4b5cfa8f9df3a6c93b3abca037142a528/03_train_gnn.py#L161), it means you only need `10000` training samples each epoch other than all the training samples to reproduce the result in the paper? 
3. In my own code, I create multiple non-communicated instance generators for parallel generation, which I have no idea whether It's safe to do that.
```
# main function for parallel generation
def parallel_generation(MAX_SAMPLES):
    global episode_counter, sample_counter, flag
    
    # We can pass custom SCIP parameters easily
    scip_parameters = {'separating/maxrounds': 0, 'presolving/maxrestarts': 0, 'limits/time': 3600}
    
    # Note how we can tuple observation functions to return complex state information
    env = ecole.environment.Branching(observation_function=(ExploreThenStrongBranch(expert_probability=flag.expert_probability), 
                                        ecole.observation.NodeBipartite()), scip_params=scip_parameters)
    
    # This will seed the environment for reproducibility
    # env.seed(0) 
    
    Path('samples/').mkdir(exist_ok=True)
    
    # We will solve problems (run episodes) until we have saved enough samples
    if flag.task == "set_cover":
        instances = ecole.instance.SetCoverGenerator(n_rows=flag.n_rows, n_cols=1000, density=0.05)
    elif flag.task == "capacitated_facility_location":
        instances = ecole.instance.CapacitatedFacilityLocationGenerator(n_customers=flag.n_customers)
    elif flag.task == "independent_set":
        instances = ecole.instance.IndependentSetGenerator(n_nodes=flag.n_nodes)
    elif flag.task == "combinatorial_auction":
        instances = ecole.instance.CombinatorialAuctionGenerator(n_items=flag.n_items, n_bids=flag.n_bids)

    max_samples_reached = False
    while not max_samples_reached:
        observation, action_set, _, done, _ = env.reset(next(instances))
        while not done:
            (scores, scores_are_expert), node_observation = observation
            action = action_set[scores[action_set].argmax()]

            # Only save samples if they are coming from the expert (strong branching)
            if scores_are_expert and not max_samples_reached:
                data = [node_observation, action, action_set, scores]
                with sample_counter.get_lock():
                    if sample_counter.value < MAX_SAMPLES:
                        sample_counter.value += 1  # make sure value is unique
                        filename = f'samples/sample_{sample_counter.value + flag.offset}.pkl'
                    else:  # collected enough samples, we finish the current episode but stop saving samples
                        max_samples_reached = True
                        break
                    
                with gzip.open(filename, 'wb') as f:
                    pickle.dump(data, f)

            observation, action_set, _, done, _ = env.step(action)
        with episode_counter.get_lock():
            episode_counter.value += 1  # make sure value is unique
            print(f"Episode {episode_counter.value}, {sample_counter.value} samples collected so far")

with mp.Pool(n_jobs, initializer=data_init, initargs=(episode_counter, sample_counter, FLAGS,)) as pool:
	pool.map(parallel_generation, repeat(MAX_SAMPLES, n_jobs))
```
I just checked most of the code and nothing special was found except the training protocol. How weird.

**Reply by dchetelat** [reply]
> Did you train the model on big instances or small instances? Since the nodes on big are less than small.

This is trained on small instances, evaluated on small (different test instances), medium and big. I'm not sure what you mean by "since the nodes on big are less than small" - on the small instances the trees have 134 nodes on average, while on big they have 64,842 nodes on average?

> From this line, it means you only need 10000 training samples each epoch other than all the training samples to reproduce the result in the paper?

Yes, this is a small subtlety not well explained in the original paper. To save training time, at every epoch we shuffle the 100k samples but only do a pass on the first 10k. Since the data is shuffled between every epoch, these end up being different 10k at every epoch. You can probably train on the full 100k at every epoch if you wish, with no negative consequences, I suppose?

> In my own code, I create multiple non-communicated instance generators for parallel generation, which I have no idea whether It's safe to do that.

Yes, I think this should be fine.

> I just checked most of the code and nothing special was found except the training protocol. How weird.

Well, again, at this point there is still the possibility that this is coming from the Ecole instance generators, so I wouldn't automatically assume something is up with your code. We'll see.

**Reply by dxyzx0** [reply]
Thank you. I mean in small instance evaluation, gnn node num 134 is larger than scip default node num 51 but still gets time benefit, which is weird in my opinion. Hence, I assume you trained on big instances, which is wrong.

**Reply by dchetelat** [reply]
Oh I see. Yeah, the same thing was happening in the [original paper](https://papers.nips.cc/paper/2019/file/d14c2267d848abeb81fd590f371d39bd-Paper.pdf) as well (Table 2 on page 7): on setcover, on 5 seeds, SCIP default got 54 nodes on average while the GNN had 134. In fact, this held for all instances classes: even though the model is trained on small instances, ultimately reliability pseudocost remains better on small in nb nodes. ¯\\\_(ツ)\_/¯

I will say, however, that in many ways the number of nodes reported by SCIP for reliability pseudocost (and non-vanilla strong branching) are arguably underestimated: there are side-effects and some nodes are "created" but automatically fathomed from the tree when branching. Whether you should consider these nodes has having been opened (hence should be counted in the tree size) is a philosophical question. On this topic you might find Gamrath and Schubert (2018) interesting, as well as the discussion in Section 4.1 of Zarpellon et al. (2020), which post-date our paper. Gamrath and Schubert propose a "fair node count" that could be used to correct that deficiency, but unfortunately it requires deactivating primal heuristics. As far as I know nobody has proposed yet a general-purpose "fair node count" that could be used in regular solving. I had opened [an issue](https://github.com/ds4dm/ecole/issues/92) about this a while back.

Gamrath, G., & Schubert, C. (2018). Measuring the impact of branching rules for mixed-integer programming. In Operations Research Proceedings 2017 (pp. 165-170). Springer
Zarpellon, G., Jo, J., Lodi, A., & Bengio, Y. (2021, May). Parameterizing Branch-and-Bound Search Trees to Learn Branching Policies. In Proceedings of the AAAI Conference on Artificial Intelligence (Vol. 35, No. 5, pp. 3931-3939).

**Reply by dxyzx0** [reply]
BTW, I found a weird phenomenon that the training/validation accuracy of **combinatorial auction** won't increase. Can you reproduce the problem? Other tasks work well.
```
[2021-10-21 20:12:46.586854] EPOCH 253...
[2021-10-21 20:13:00.615919] TRAIN LOSS: 2.607  acc@1: 0.474 acc@3: 0.708 acc@5: 0.816 acc@10: 0.934
[2021-10-21 20:13:16.086814] VALID LOSS: 2.611  acc@1: 0.474 acc@3: 0.703 acc@5: 0.812 acc@10: 0.931
[2021-10-21 20:13:16.087240] EPOCH 254...
[2021-10-21 20:13:29.891938] TRAIN LOSS: 2.604  acc@1: 0.474 acc@3: 0.711 acc@5: 0.818 acc@10: 0.933
[2021-10-21 20:13:45.182254] VALID LOSS: 2.611  acc@1: 0.474 acc@3: 0.703 acc@5: 0.812 acc@10: 0.930
[2021-10-21 20:13:45.182633] EPOCH 255...
[2021-10-21 20:13:58.630930] TRAIN LOSS: 2.593  acc@1: 0.473 acc@3: 0.703 acc@5: 0.815 acc@10: 0.930
[2021-10-21 20:14:13.781607] VALID LOSS: 2.611  acc@1: 0.473 acc@3: 0.703 acc@5: 0.812 acc@10: 0.931
[2021-10-21 20:14:13.782002] EPOCH 256...
[2021-10-21 20:14:27.094316] TRAIN LOSS: 2.591  acc@1: 0.478 acc@3: 0.706 acc@5: 0.816 acc@10: 0.934
[2021-10-21 20:14:42.387882] VALID LOSS: 2.611  acc@1: 0.473 acc@3: 0.703 acc@5: 0.812 acc@10: 0.930
[2021-10-21 20:14:42.388286] EPOCH 257...
[2021-10-21 20:14:55.826369] TRAIN LOSS: 2.579  acc@1: 0.480 acc@3: 0.708 acc@5: 0.817 acc@10: 0.935
[2021-10-21 20:15:10.879072] VALID LOSS: 2.611  acc@1: 0.474 acc@3: 0.703 acc@5: 0.812 acc@10: 0.930
[2021-10-21 20:15:10.879466] EPOCH 258...
[2021-10-21 20:15:24.284918] TRAIN LOSS: 2.575  acc@1: 0.473 acc@3: 0.716 acc@5: 0.820 acc@10: 0.939
[2021-10-21 20:15:39.623349] VALID LOSS: 2.612  acc@1: 0.475 acc@3: 0.703 acc@5: 0.812 acc@10: 0.930
[2021-10-21 20:15:39.623798] EPOCH 259...
[2021-10-21 20:15:53.617875] TRAIN LOSS: 2.597  acc@1: 0.475 acc@3: 0.707 acc@5: 0.812 acc@10: 0.935
[2021-10-21 20:16:09.005999] VALID LOSS: 2.611  acc@1: 0.474 acc@3: 0.703 acc@5: 0.812 acc@10: 0.930
[2021-10-21 20:16:09.006399] EPOCH 260...
[2021-10-21 20:16:22.500397] TRAIN LOSS: 2.576  acc@1: 0.461 acc@3: 0.705 acc@5: 0.810 acc@10: 0.935
[2021-10-21 20:16:38.666407] VALID LOSS: 2.612  acc@1: 0.475 acc@3: 0.704 acc@5: 0.812 acc@10: 0.930
[2021-10-21 20:16:38.666809] EPOCH 261...
[2021-10-21 20:16:52.327668] TRAIN LOSS: 2.598  acc@1: 0.469 acc@3: 0.703 acc@5: 0.818 acc@10: 0.932
[2021-10-21 20:17:06.524168] VALID LOSS: 2.611  acc@1: 0.475 acc@3: 0.704 acc@5: 0.812 acc@10: 0.930
[2021-10-21 20:17:06.524482] EPOCH 262...
[2021-10-21 20:17:19.644947] TRAIN LOSS: 2.595  acc@1: 0.485 acc@3: 0.711 acc@5: 0.818 acc@10: 0.934
[2021-10-21 20:17:35.656261] VALID LOSS: 2.611  acc@1: 0.474 acc@3: 0.703 acc@5: 0.812 acc@10: 0.930
[2021-10-21 20:17:35.656523]   30 epochs without improvement, early stopping
[2021-10-21 20:17:50.041247] BEST VALID LOSS: 2.611  acc@1: 0.474 acc@3: 0.704 acc@5: 0.812 acc@10: 0.931
```

**Reply by dchetelat** [reply]
Sorry, you're not giving me enough details. Is this from the new repo I created? Where did the instances come from?
What do you mean by "the other tasks work well"?

**Comment by dxyzx0** [comment]
The training accuracy doesn't increase as you can see in the picture. This situation only happens for combinatorial auction, not other 3 tasks.
----------

该邮件从移动设备发送


--------------原始邮件--------------
发件人："Didier Chételat ***@***.***&gt;;
发送时间：2021年10月22日(星期五) 晚上9:34
收件人："ds4dm/ecole" ***@***.***&gt;;
抄送："dxyzx ***@***.***&gt;;"Mention ***@***.***&gt;;
主题：Re: [ds4dm/ecole] Cannot reproduce the experiment result in the paper (Discussion #262)
-----------------------------------

 

 
Sorry, you're not giving me enough details. Is this from the new repo I created? Where did the instances come from?
 What do you mean by "the other tasks work well"?
 
—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.

**Reply by dchetelat** [reply]
Hi, I understand. But I cannot help you if you do not answer my questions.

1. Is this from the new repo I created?
2. Where did the instances come from?

**Reply by dxyzx0** [reply]
1. No I still use ecole.
2. The instances come from the InstanceGenerator for combinatorial auction.

**Reply by dchetelat** [reply]
Hi, I reproduced this issue on my side as well - the combinatorial auctions instances are different (at minimum, harder) than what were used in the original paper. So this seems to be an issue about instance generators. I'll look into it.
